aws/resource_aws_cloudsearch_service.go

```diff
@@ -0,0 +1,471 @@
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"search": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"facet": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+						"return": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"sort": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+						"highlight": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+						"analysis_scheme": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"default_value": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"access_policy": {
+				Type:             schema.TypeString,
+				ValidateFunc:     validateIAMPolicyJson,
+				Required:         true,
+				DiffSuppressFunc: suppressEquivalentAwsPolicyDiffs,
+			},
+		},
+	}
+}
+
+func resourceAwsCloudSearchDomainCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cloudsearchconn
+
+	input := cloudsearch.CreateDomainInput{
+		DomainName: aws.String(d.Get("domain_name").(string)),
+	}
+
+	output, err := conn.CreateDomain(&input)
+	if err != nil {
+		return fmt.Errorf("%s %q", err, d.Get("domain_name").(string))
+	}
+
+	d.SetId(*output.DomainStatus.ARN)
+	err = resourceAwsCloudSearchDomainUpdate(d, meta)
+
+	return err
+}
+
+func resourceAwsCloudSearchDomainRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cloudsearchconn
+
+	input := cloudsearch.DescribeIndexFieldsInput{
+		DomainName: aws.String(d.Get("domain_name").(string)),
+	}
+
+	_, err := conn.DescribeIndexFields(&input)
+
+	return err
+}
+
+func resourceAwsCloudSearchDomainUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cloudsearchconn
+
+	err := updateScalingParameters(d, meta, conn)
+	if err != nil {
+		return err
+	}
+
+	updated, err := defineIndexFields(d, meta, conn)
+	if err != nil {
+		return err
+	}
+
+	err = updateAccessPolicy(d, meta, conn)
+	if err != nil {
+		return err
+	}
+
+	if updated {
+		_, err := conn.IndexDocuments(&cloudsearch.IndexDocumentsInput{
+			DomainName: aws.String(d.Get("domain_name").(string)),
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func updateScalingParameters(d *schema.ResourceData, meta interface{}, conn *cloudsearch.CloudSearch) error {
+	input := cloudsearch.UpdateScalingParametersInput{
+		DomainName: aws.String(d.Get("domain_name").(string)),
+		ScalingParameters: &cloudsearch.ScalingParameters{
+			DesiredInstanceType:     aws.String(d.Get("instance_type").(string)),
+			DesiredPartitionCount:   aws.Int64(int64(d.Get("partition_count").(int))),
```

Comment:
As per http://docs.aws.amazon.com/cloudsearch/latest/developerguide/API_ScalingParameters.html, it is valid only if the instance is `m3.2xlarge`

> DesiredPartitionCount
> The number of partitions you want to preconfigure for your domain. Only valid when you select m3.2xlarge as the instance type.